core/drivers/clk/clk-stm32mp13.c

```diff
@@ -0,0 +1,1727 @@
+
+	return 0;
+}
+
+static int
+stm32_clk_oscillators_wait_lse_ready(__maybe_unused struct clk_stm32_priv *priv,
+				     struct stm32_clk_platdata *pdata)
+{
+	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(OSC_LSE);
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSE];
+
+	if (osci->freq != 0U)
+		clk_stm32_wait_ready_gate(osc_data->gate_id, true);
+
+	return 0;
+}
+
+static void stm32_clk_oscillators_enable(struct clk_stm32_priv *priv,
+					 struct stm32_clk_platdata *pdata)
+{
+	stm32_enable_oscillator_hse(priv, pdata);
+	stm32_enable_oscillator_lse(priv, pdata);
+	stm32_enable_oscillator_lsi(priv, pdata);
+	stm32_enable_oscillator_csi(priv, pdata);
+}
+
+enum stm32_pll_id {
+	PLL1_ID,
+	PLL2_ID,
+	PLL3_ID,
+	PLL4_ID,
+	PLL_NB
+};
+
+enum stm32mp1_plltype {
+	PLL_800,
+	PLL_1600,
+	PLL_2000,
+	PLL_TYPE_NB
+};
+
+#define RCC_OFFSET_PLLXCR		0
+#define RCC_OFFSET_PLLXCFGR1		4
+#define RCC_OFFSET_PLLXCFGR2		8
+#define RCC_OFFSET_PLLXFRACR		12
+#define RCC_OFFSET_PLLXCSGR		16
+
+struct stm32_clk_pll {
+	enum stm32mp1_plltype plltype;
+	uint16_t gate_id;
+	uint16_t mux_id;
+	uint16_t reg_pllxcr;
+};
+
+struct stm32mp1_pll {
+	uint8_t refclk_min;
+	uint8_t refclk_max;
+};
+
+/* Define characteristic of PLL according type */
+static const struct stm32mp1_pll stm32mp1_pll[PLL_TYPE_NB] = {
+	[PLL_800] = {
+		.refclk_min = 4,
+		.refclk_max = 16,
+	},
+	[PLL_1600] = {
+		.refclk_min = 8,
+		.refclk_max = 16,
+	},
+	[PLL_2000] = {
+		.refclk_min = 8,
+		.refclk_max = 16,
+	}
+};
+
+#define CLK_PLL_CFG(_idx, _type, _gate_id, _mux_id, _reg)\
+	[(_idx)] = {\
+		.gate_id = (_gate_id),\
+		.mux_id = (_mux_id),\
+		.plltype = (_type),\
+		.reg_pllxcr = (_reg),\
+	}
+
+static const struct stm32_clk_pll stm32_mp13_clk_pll[PLL_NB] = {
+	CLK_PLL_CFG(PLL1_ID, PLL_2000, GATE_PLL1, MUX_PLL12, RCC_PLL1CR),
+	CLK_PLL_CFG(PLL2_ID, PLL_1600, GATE_PLL2, MUX_PLL12, RCC_PLL2CR),
+	CLK_PLL_CFG(PLL3_ID, PLL_800, GATE_PLL3, MUX_PLL3, RCC_PLL3CR),
+	CLK_PLL_CFG(PLL4_ID, PLL_800, GATE_PLL4, MUX_PLL4, RCC_PLL4CR),
+};
+
+static const struct stm32_clk_pll *clk_stm32_pll_data(unsigned int idx)
+{
+	return &stm32_mp13_clk_pll[idx];
+}
+
+/* Clock TREE configuration */
+
+static unsigned int stm32_clk_configure_clk_get_binding_id(uint32_t data)
+{
+	return (data & CLK_ID_MASK) >> CLK_ID_SHIFT;
+}
+
+static int stm32_clk_configure_clk(__maybe_unused struct clk_stm32_priv *priv,
+				   uint32_t data)
+{
+	int sel = (data & CLK_SEL_MASK) >> CLK_SEL_SHIFT;
+	int enable = (data & CLK_ON_MASK) >> CLK_ON_SHIFT;
+	int clk_id = 0;
+	int ret = 0;
+	int mux = -1;
+	int gate = -1;
+
+	clk_id = stm32_clk_configure_clk_get_binding_id(data);
+
+	switch (clk_id)	{
+	case CK_MCO1:
+		mux = MUX_MCO1;
+		gate = GATE_MCO1;
+		break;
+
+	case CK_MCO2:
+		mux = MUX_MCO2;
+		gate = GATE_MCO2;
+		break;
+	default:
+		ret = -1;
+		break;
+	}
+
+	if (ret != 0)
+		return ret;
+
+	clk_stm32_set_parent_mux(mux, sel);
+
+	if (enable)
+		clk_stm32_enable_gate(gate);
```

Comment:
maybe call straight `clk_stm32_endisable_gate(gate, true);` to not bother with return value.